File: ruby/medium/construct_binary_tree_from_preorder_and_inorder_traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {Integer[]} preorder
# @param {Integer[]} inorder
# @return {TreeNode}
def build_tree(preorder, inorder)
    map = {}
    root = nil
    inorder.each_with_index do |n, i|
        map[n] = i
    end
    
    preorder.each do |n|
        root = insert(root, n, map)
    end
    return root
end

def insert(root, n, map)
    return TreeNode.new(n) if root.nil?
    index = map[n]
    root_index = map[root.val]
    if(index < root_index)
        root.left = insert(root.left, n, map)
    else
        root.right = insert(root.right, n, map)
    end
    return root
end

File: javascript/medium/constructBinaryTreeFromPreorderAndInorderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
    const map = new Map();
    let root = null;
    for(let i = 0; i < inorder.length; i++) {
        map.set(inorder[i], i);
    }
    
    for(let i = 0; i < preorder.length; i++) {
        root = insert(root, preorder[i], map);
    }
    return root;
};

var insert = function(root, num, map) {
    if(root === null) {
        root = new TreeNode(num);
    } else {
        if(map.get(num) < map.get(root.val)) {
            root.left = insert(root.left, num, map);
        } else {
            root.right = insert(root.right, num, map);
        }
    }
    return root;
};
